drivers/dma/dma_mcux_edma.c

```diff
@@ -164,19 +166,15 @@ static void channel_irq(edma_handle_t *handle)
 		handle->header = (int8_t)new_header;
 		handle->tcdUsed -= (int8_t)tcds_done;
 
-		/* Clear the done flag before the callback so a user can optionally
-		 * install (reload) the next handle during the callback if desired
-		 */
-		if (transfer_done) {
-			handle->base->CDNE = handle->channel;
-		}
-
 		/* Invoke callback function. */
 		if (handle->callback != NULL) {
 			(handle->callback)(handle, handle->userData,
 					   transfer_done, tcds_done);
 		}
 
+		if (transfer_done) {
+			handle->base->CDNE = handle->channel;
```

Comment:
I went back through and figured out why it was an issue to call `EDMA_SubmitTransfer()` from within the callback ISR context.

If you look at `fsl_edma.c`, considering the case of a ping-pong buffer, we end up in this block:
https://github.com/NXPmicro/mcux-sdk/blob/79cdc5799191a7411479a2617e28a191f4383276/drivers/edma/fsl_edma.c#L1176-L1207

In this block, we've correctly loaded the next DMA descriptor to point to, and now we are trying to set the ESG bit for the active DMA descriptor to enable it to continue on into the reloaded buffer. However, because we haven't cleared the DONE bit yet, writes to ESG are disabled and so it looks to the driver like the previous transfer finished (see the comment [here](https://github.com/NXPmicro/mcux-sdk/blob/main/drivers/edma/fsl_edma.c#L1209)).

I get around this in the PR that I pushed where I just delay reloading the DMA reload until we're out of ISR context in the case that `dma_reload` is called from w/in ISR, but this feels a bit ugly to me. Happy to take suggestions if you have thoughts about a better way to handle this.

Reply:
Sounds good! I think I've got this correctly updated to use the changes from the hal_nxp PR ([#144](https://github.com/zephyrproject-rtos/hal_nxp/pull/144/)), and to use the `HandleIRQ` function from within the driver itself instead of the custom `channel_irq`. I tested with the driver fixes on my board and all seems to work as intended.